Replace manually maintained code about frameworks with generated code.

rolfbjarne opened the pull request:
* [src] Generate the string constants for each framework from our list of
  frameworks.

  This makes it less error prone to add new frameworks (one less place to add 
  them), and easier to add new platforms (only one place to make sure the
  list of frameworks is correct).

  This also revelead a few issues in our list of frameworks (missing
  frameworks, frameworks that have been moved, etc.), which have been fixed.

* [mmp] Generate the library path -> namespace map for the linker to optimize away dlopen statements.

monojenkins commented:
Build failure
❌ [Build failed or was aborted](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15986/)

✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15986//console)
🔥 [Build failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/15986//console) 🔥

spouliot commented on src/Makefile:
```diff
@@ -725,6 +728,7 @@ WATCHOS_WARNINGS_TO_FIX = -nowarn:219,618,114,414,1635,3021,$(IOS_WARNINGS_THAT_
 
 WATCHOS_EXTRA_CORE_SOURCES = \
     $(WATCH_BUILD_DIR)/Constants.cs    \
+	$(BUILD_DIR)/Constants.watchos.generated.cs \
```
indent ?

spouliot commented on src/generate-frameworks-constants/Program.cs:
```diff
@@ -0,0 +1,56 @@
+
+using Xamarin.Utils;
+
+namespace GenerateTypeForwarders {
```
^ weird name ;-)

rolfbjarne replied:
c&p...

spouliot commented on src/generate-frameworks-constants/generate-frameworks-constants.csproj:
```diff
@@ -0,0 +1,48 @@
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Mono.Cecil" Version="0.11.3" />
```
did not see Cecil used

rolfbjarne replied:
Fixed.

spouliot reviewed (commented):
❤️  it

maybe `src/generate-frameworks-constants` should be under `tools` ? that's where I would look for it instinctively

rolfbjarne commented:
> ❤️ it
> 
> maybe `src/generate-frameworks-constants` should be under `tools` ? that's where I would look for it instinctively

The problem is that tools/ is built after src/:

https://github.com/xamarin/xamarin-macios/blob/14ad1a122bd1bacae402dc4a2337eed438cd2a71/Makefile#L2

So yes, I could move it to tools/, but I would still have to build it from src/ (something like `make -C $(TOP)/tools/generate-frameworks-constants`), and that's somewhat ugly too, so I decided to just put this in src/ (but either way would work fine for me).

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16018//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16018//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16018//API_20diff_20_28from_20stable_29)
🔥 [Failed to compare API and create generator diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16018//console) 🔥
&nbsp;&nbsp;&nbsp;&nbsp;Failed to build src/
&nbsp;&nbsp;&nbsp;&nbsp;Search for `Comparing API & creating generator diff` in the log to view the complete log.
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16018//Test_20Report) 🔥

# Test results

<details>
<summary>2 tests failed, 78 tests passed.</summary>

## Failed tests

 * apitest/Mac Modern/Debug: Failed (Tests run: 258 Passed: 255 Inconclusive: 1 Failed: 1 Ignored: 2)
 * apitest/Mac Full/Debug: Failed (Tests run: 258 Passed: 255 Inconclusive: 1 Failed: 1 Ignored: 2)
</details>

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16036//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16036//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16036//API_20diff_20_28from_20stable_29)
🔥 [Failed to compare API and create generator diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16036//console) 🔥
&nbsp;&nbsp;&nbsp;&nbsp;** Error: The following files were modified, and they shouldn't have been:
&nbsp;&nbsp;&nbsp;&nbsp;/Users/builder/jenkins/workspace/xamarin-macios-pr-builder/src/generate-type-forwarders/obj/Debug/generate-type-forwarders.csproj.CopyComplete
&nbsp;&nbsp;&nbsp;&nbsp;Search for `Comparing API & creating generator diff` in the log to view the complete log.
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16036//Test_20Report) 🔥

# Test results

<details>
<summary>1 tests failed, 79 tests passed.</summary>

## Failed tests

 * link all/tvOS - simulator/Debug: Failed
</details>

monojenkins commented:
Build success
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16055//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16055//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16055//API_20diff_20_28from_20stable_29)
⚠️ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16055//API_20diff_20_28PR_20only_29) (🔥 breaking changes 🔥)
ℹ️ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16055//Generator_20Diff) (please review changes)
✅ [Test run succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16055//Test_20Report)

rolfbjarne commented:
The breaking API changes are expected:

- public const string CFNetworkLibrary = ~"/System/Library/Frameworks/CoreServices.framework/Frameworks/CFNetwork.framework/CFNetwork"~ "/System/Library/Frameworks/CFNetwork.framework/CFNetwork";

Unsure if this was a mistake or if CFNetwork was a part of CoreServices at some point, but in any case CFNetwork is not a subframework of CoreServices anymore (in 10.15).

- public const string CallKitLibrary = ~"/System/Library/Frameworks/CallKitLibrary.framework/CallKitLibrary"~ "/System/Library/Frameworks/CallKit.framework/CallKit";

The previous value was just wrong.

- public const string CoreBluetoothLibrary = ~"/System/Library/Frameworks/IOBluetooth.framework/Versions/A/Frameworks/CoreBluetooth.framework/CoreBluetooth"~ "/System/Library/Frameworks/CoreBluetooth.framework/CoreBluetooth";

Moved from inside an umbrella framework (IOBluetooth) into its own framework in 10.10

- public const string CoreImageLibrary = ~"/System/Library/Frameworks/QuartzCore.framework/Versions/A/Frameworks/CoreImage.framework/CoreImage"~ "/System/Library/Frameworks/CoreImage.framework/CoreImage";

Moved from inside an umbrella framework (QuartzCore) into its own framework in 10.11

- public const string CoreTextLibrary = ~"/System/Library/Frameworks/ApplicationServices.framework/Frameworks/CoreText.framework/CoreText"~ "/System/Library/Frameworks/CoreText.framework/CoreText";

It's own framework since at least 10.9

- public const string ImageIOLibrary = ~"/System/Library/Frameworks/ApplicationServices.framework/Frameworks/ImageIO.framework/ImageIO"~ "/System/Library/Frameworks/ImageIO.framework/ImageIO";

It's own framework since at least 10.9

- public const string ImageKitLibrary = ~"/System/Library/Frameworks/Quartz.framework/Frameworks/ImageKit.framework/ImageKit"~ "/System/Library/Frameworks/Quartz.framework/Versions/A/Frameworks/ImageKit.framework/ImageKit";

The same path, because of symlinks (now follows the same pattern as other paths).

- public const string PdfKitLibrary = ~"/System/Library/Frameworks/Quartz.framework/Frameworks/PDFKit.framework/PDFKit"~ "/System/Library/Frameworks/Quartz.framework/Versions/A/Frameworks/PDFKit.framework/PDFKit";

The same path, because of symlinks (now follows the same pattern as other paths).

- public const string QuartzComposerLibrary = ~"/System/Library/Frameworks/Quartz.framework/Frameworks/QuartzComposer.framework/QuartzComposer"~ "/System/Library/Frameworks/Quartz.framework/Versions/A/Frameworks/QuartzComposer.framework/QuartzComposer";

The same path, because of symlinks (now follows the same pattern as other paths).

- public const string QuickLookUILibrary = ~"/System/Library/Frameworks/Quartz.framework/Frameworks/QuickLookUI.framework/QuickLookUI"~ "/System/Library/Frameworks/Quartz.framework/Versions/A/Frameworks/QuickLookUI.framework/QuickLookUI";

The same path, because of symlinks (now follows the same pattern as other paths).

- public const string SearchKitLibrary = ~"/System/Library/Frameworks/CoreServices.framework/Frameworks/SearchKit.framework/SearchKit"~ "/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/SearchKit.framework/SearchKit";

The same path, because of symlinks (now follows the same pattern as other paths).

monojenkins commented:
Build failure
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16056//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16056//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16056//API_20diff_20_28from_20stable_29)
⚠️ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16056//API_20diff_20_28PR_20only_29) (🔥 breaking changes 🔥)
ℹ️ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16056//Generator_20Diff) (please review changes)
🔥 [Test run failed](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16056//Test_20Report) 🔥

# Test results

<details>
<summary>19 tests failed, 61 tests passed.</summary>

## Failed tests

 * monotouch-test/iOS Unified 64-bits - simulator/Debug: TimedOut
 * monotouch-test/iOS Unified 64-bits - simulator/Debug (LinkSdk): TimedOut
 * monotouch-test/iOS Unified 64-bits - simulator/Debug (static registrar): TimedOut
 * monotouch-test/iOS Unified 64-bits - simulator/Release (all optimizations): TimedOut
 * fsharp/iOS Unified 64-bits - simulator/Debug: TimedOut
 * framework-test/iOS Unified 64-bits - simulator/Debug: TimedOut
 * xcframework-test/iOS Unified 64-bits - simulator/Debug: TimedOut
 * interdependent-binding-projects/iOS Unified 64-bits - simulator/Debug: TimedOut
 * introspection/iOS Unified 64-bits - simulator/Debug: TimedOut
 * dont link/iOS Unified 64-bits - simulator/Debug: TimedOut
 * dont link/iOS Unified 64-bits - simulator/Release: TimedOut
 * link all/iOS Unified 64-bits - simulator/Debug: TimedOut
 * link all/iOS Unified 64-bits - simulator/Release: TimedOut
 * link sdk/iOS Unified 64-bits - simulator/Debug: TimedOut
 * link sdk/iOS Unified 64-bits - simulator/Release: TimedOut
 * mono-native-compat/iOS Unified 64-bits - simulator/Debug: TimedOut
 * mono-native-compat/iOS Unified Today Extension 64-bits - simulator/Debug: TimedOut
 * mono-native-unified/iOS Unified 64-bits - simulator/Debug: TimedOut
 * mono-native-unified/iOS Unified Today Extension 64-bits - simulator/Debug: TimedOut
</details>

rolfbjarne commented:
build

monojenkins commented:
Build success
✅ [Provisioning succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16058//console)
✅ [Build succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16058//console)
✅ [API Diff (from stable)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16058//API_20diff_20_28from_20stable_29)
⚠️ [API Diff (from PR only)](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16058//API_20diff_20_28PR_20only_29) (🔥 breaking changes 🔥)
ℹ️ [Generator Diff](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16058//Generator_20Diff) (please review changes)
✅ [Test run succeeded](https://jenkins.mono-project.com/job/xamarin-macios-pr-builder/16058//Test_20Report)

dalexsoto reviewed (approved):
Nice!
